src/main.js

```diff
@@ -1,144 +1,175 @@
-let knownRules
-
-const loadDeps = () => {
-  if (!path) {
-    path = require('path')
-  }
-  if (!worker) {
-    worker = require('./worker-manager')
-  }
-  if (!configInspector) {
-    configInspector = require('./eslint-config-inspector')
-  }
-  if (!debug) {
-    debug = require('./debug')
-  }
-  if (!linterMessage) {
-    linterMessage = require('./linter-message')
-  }
-  if (!knownRules) {
-    knownRules = require('./rules').default
-  }
-}
+// NOTE: Requiring files adds significant time to package startup. To avoid this
+// delay, we are *not* requiring at the top of this file. Use idleCallbacks to
+// attempt to pre-cache requires for fast resolution if scheduler allows. Since
+// all requires are statically defined and inside some function, Atom should be
+// able "snapshot" the package.
+//
+const dependencies = [
+  () => require('./validate/editor'),
+  () => require('./linter-message'),
+  () => require('./worker-manager'),
+  () => require('./eslint-config-inspector'),
+  () => require('./debug'),
+  () => require('./rules'),
+  () => require('path'),
+  () => require('./atom-config'),
+]
+
+// Miscellaneous idle tasks.
+//
+const idleTasks = [
+  // Pre-start the worker
+  () => require('./worker-manager').task.start(),
+  // Install peer packages
+  () => require('atom-package-deps').install('linter-eslint'),
+]
+
+// Run idle tasks with all reasonable attempts to avoid blocking.
+//
+const makeIdleRunner = () => {
+  const { requestIdleCallback, cancelIdleCallback } = window
+  let id
+  let tasks
+
+  const loadLazily = (deadline) => {
+    // While time alotted by scheduler and work remaining
+    while (deadline.timeRemaining() && tasks.length) {
+      // Remove and run function from end of list.
+      tasks.pop()()
+    }
 
-const makeIdleCallback = (work) => {
-  let callbackId
-  const callBack = () => {
-    idleCallbacks.delete(callbackId)
-    work()
+    // If work remains, request more time
+    if (tasks.length) id = requestIdleCallback(loadLazily)
   }
-  callbackId = window.requestIdleCallback(callBack)
-  idleCallbacks.add(callbackId)
-}
 
-const scheduleIdleTasks = () => {
-  const linterEslintInstallPeerPackages = () => {
-    require('atom-package-deps').install('linter-eslint')
-  }
-  const linterEslintLoadDependencies = loadDeps
-  const linterEslintStartWorker = () => {
-    loadDeps()
-    worker.task.start()
+  return (requestedTasks) => {
+    // Spread avoids external mutation and allows non-array iterable
+    tasks = [...requestedTasks]
+    // Init first request
+    id = requestIdleCallback(loadLazily)
+    // Return a disposable
+    return { dispose: () => cancelIdleCallback(id) }
   }
+}
 
-  if (!atom.inSpecMode()) {
-    makeIdleCallback(linterEslintInstallPeerPackages)
-    makeIdleCallback(linterEslintLoadDependencies)
-    makeIdleCallback(linterEslintStartWorker)
+// // Simplified composite disposable avoids requiring atom core in main
+// //
+const makeCompositeDisposable = () => {
```

Comment:
What is your reasoning for needing to write this? In chat you briefly mentioned it was "faster", but I'd have to see some pretty big performance gains from this to justify writing our own version instead of just using what Atom already has written and maintains for us.

Reply:
I came up with it when converting to all `require`s instead of `import`s. 

Of course that plan has been scrapped, but when using require, it's easier to notice that in order to import `CompositeDisposable` the entire `atom` entry point with tons of classes we have no use for must be parsed through to dig out the class we want.

I had just recently before that worked out what that class was actually doing (basically just holding a list of things with a `dispose` method and calling said method for everyone in the list). So I threw together a function that provided the same basic behavior just to see.

I don't know what you're perfs will show when you manage to zoom in on our startup times. But I put a lot of tries into testing this across varying loads, with only Atom running or with tons of apps open consuming my CPU. 

Except for some very occasional and totally random outliers that went the other direction, this change under all cases (at least on my system) results in a minimum 6% speed improvement for running the full test suite, usually closer to 10%. 

So for example, if my CPU load was finishing the suite in 80s, I could make the tiny change of swapping this in for the class import and finish the test suite in 72s under the same load, pretty consistently.

Reply:
> it's easier to notice that in order to import CompositeDisposable the entire atom entry point with tons of classes we have no use for must be parsed through to dig out the class we want.

`atom` is embedded in the VM, the only cost to `require`ing it is resolving the package name. You can see this yourself in the profile. If there is any difference in the performance it's in the implementation of the `CompositeDisposable` itself, which would make me question what is so different that it is that much faster.

Reply:
Please go ahead and pull this to your machine and do some of your own A/B testing and check it out in perfs. If there is no cost to the import then I really don't know where the performance is coming from. 

Maybe it is a fluke on my machine. Maybe `CompositeDisposable` is overly complicated for its job, or poorly written, or written to handle a bunch of edge cases we don't care about. All I know at this point is I saw significant and consistent speedup of the test suite that directly correlates with this change.